### PR TITLE
expression: fix wrong caseWhen function for enum type

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/29357

Problem Summary:

### What is changed and how it works?
There are two problem in this issue:
Problem 1:
1. caseWhenFunctionClass.getFunction get fieldType Enum from children, and construct scalarFunction caseWhenString.
2. compareFunctionClass.getFunction get fieldType Int and Enum, and set EnumSetAsInt flag, but not re-construct caseWhen function.
3. Get the final scalarFunction EQInt(1, CaseWhenString(0, Enum)) and pushdown.  CaseWhenString can't get int value.

Problem 2:
```
[tidb]> SELECT 5 = (case when 0 <=> 1 then a else a end) from t;
ERROR 1105 (HY000): baseBuiltinFunc.vecEvalInt() should never be called, please contact the TiDB team for help
```
1. BuildCastFunction build a cast CastStringAsString, and the fieldType is Enum.
2. compareFunctionClass.getFunction get fieldType Int and Enum, and set EnumSetAsInt flag, but not re-construct Cast function.
3. Get the final scalarFunction EQInt(1, CastStringAsString(Enum))). CastStringAsString can't get int value.
BuildCastFunction is useless. See constant_fold.goL129. We should have the same behavior for all args.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix panic for caseWhen function with enum type
```

### CODE REVIEW

File: expression/constant_fold.go [hunk 1]
```diff
@@ -143,7 +143,7 @@ func caseWhenHandler(expr *ScalarFunction) (Expression, bool) {
 			foldedExpr.GetType().Decimal = expr.GetType().Decimal
 			return foldedExpr, isDeferredConst
 		}
-		return BuildCastFunction(expr.GetCtx(), foldedExpr, foldedExpr.GetType()), isDeferredConst
+		return foldedExpr, isDeferredConst
```
Comment (winoros) [inline expression/constant_fold.go:146]: This change is safe?

**Review by lzmhhh123 — APPROVED** [review]

**Review by ichn-hu — APPROVED** [review]

**Pull request merged**